VTF Monitoring Sample (#30326)
nrfconnect/sdk-nrf: samples/vtf_monitoring/README.rst

The :ref:`vtf_monitoring` subsystem updates the snapshots on the application core and stores them in a dedicated memory region that the Wi-Fi core can read.
This sample runs on the application core and periodically logs the stored channel values.

Requirements
************

The sample supports the following development kits:

.. table-from-sample-yaml::

Overview
********

The sample periodically logs the following :ref:`vtf_monitoring` channels:

* Die temperature - Supplied by the subsystem built-in provider, which reads the sensor selected through the ``nordic,vtf-die-temp-sensor`` devicetree chosen property.
  The nRF7120 DK board overlay points this property to the SoC internal die temperature sensor.
* Battery voltage - In the default configuration, the :kconfig:option:`CONFIG_VTF_BATTERY_VOLTAGE_MONITOR` Kconfig option is enabled.
  Live capture for battery voltage is not yet implemented in the subsystem, so the snapshot holds the default value of the Kconfig option.
* XO frequency offset - Since the :kconfig:option:`CONFIG_VTF_FREQ_OFFSET_MONITOR` Kconfig option is disabled, the snapshot holds the default value of the Kconfig option.

> The sample includes a mock provider in the :file:`src/mock_battery_provider.c` file to demonstrate a custom channel monitoring backend. It registers a channel with :c:macro:`VTF_CHANNEL_DEFINE` and simulates a battery discharging from 4200 mV to 3300 mV.

Reviewer: ```suggestion
The sample includes a mock provider in the :file:`src/mock_battery_provider.c` file to demonstrate a custom channel monitoring backend.
It registers a channel with :c:macro:`VTF_CHANNEL_DEFINE` and simulates a battery discharging from 4200 mV to 3300 mV.
```

.. _vtf_monitoring_sample:

VTF monitoring
##############

.. contents::
   :local:
   :depth: 2

> The VTF monitoring sample demonstrates how voltage, temperature, and frequency offset (VTF) data is captured and stored for the nRF Wi-Fi® subsystem.

Reviewer 2: If possible please use slightly more descriptive terms here (esp. for frequency offset which is clearly not easy to see what it means unless reader goes to the overview section)

Author: Is this not covered by the subsys docs?  `nrf/doc/nrf/libraries/others/vtf_monitoring.rst`

Reviewer 2: Well, it helps a bit when you look into the sample README and understand what it talks about without requiring to jump to the sub-system documentation. Sample README - sometimes - is all what people look at

> On the SoC, the Wi-Fi core uses this data to decide when to perform recalibration.

Reviewer: Which Soc?
```suggestion
On the SoC, the Wi-Fi core uses this data to determine when recalibration is required.
```

Author: Removed "On the SoC" so that this remains true as future board support is added.

The :ref:`vtf_monitoring` subsystem updates the snapshots on the application core and stores them in a dedicated memory region that the Wi-Fi core can read.
This sample runs on the application core and periodically logs the stored channel values.

Requirements
************

> The sample supports the following development kit:

Reviewer 2: ```suggestion
The sample supports the following development kits:
```

Reviewer: isn't there only one board as of now?
<img width="950" height="124" alt="image" src="https://github.com/user-attachments/assets/c494d28b-7644-4a92-93b5-7267bc3a0d47" />

Author: Good point - have reverted back to kit and can adjust as and when new boards are added.

.. table-from-sample-yaml::

Overview
********

The sample periodically logs the following :ref:`vtf_monitoring` channels:

* Die temperature - Supplied by the subsystem built-in provider, which reads the sensor selected through the ``nordic,vtf-die-temp-sensor`` devicetree chosen property.

>   The nRF7120 DK board overlay points this property to the SoC internal die temperature sensor.

Reviewer 2: Does this now mean that the description becomes specific to nRF7120 DK? Wonder if the overview text can be more abstract, to account for extending the sample to more platforms in future releases.

Reviewer 2: Alternatively, explain that the overview considers the nRF7120 DK as the reference development board.

Author: Change is dependent of question for @krish2718 so waiting for this to be resolved before editing - but will take into consideration before merge.

Author: 7120 ref dropped here

> * Battery voltage - In the default configuration, :kconfig:option:`CONFIG_VTF_BATTERY_VOLTAGE_MONITOR` is enabled.

Reviewer: ```suggestion
* Battery voltage - In the default configuration, the :kconfig:option:`CONFIG_VTF_BATTERY_VOLTAGE_MONITOR` Kconfig option is enabled.
```

>   Live capture for battery voltage is not yet implemented in the subsystem, so the snapshot holds the Kconfig default value.

Reviewer: ```suggestion
  Live capture for battery voltage is not yet implemented in the subsystem, so the snapshot holds the default value of the Kconfig option.
```

> * XO frequency offset - :kconfig:option:`CONFIG_VTF_FREQ_OFFSET_MONITOR` is disabled, so the snapshot holds the Kconfig default value.

Reviewer: ```suggestion
* XO frequency offset - Since the :kconfig:option:`CONFIG_VTF_FREQ_OFFSET_MONITOR` Kconfig option is disabled, the snapshot holds the default value of the Kconfig option.
```

> You can replace the battery voltage backend with a custom implementation.

Reviewer 2: Maybe keep third-person throughout the entire documentation page, for style consistency.

> The sample includes a mock provider in :file:`src/mock_battery_provider.c` that registers a channel with :c:macro:`VTF_CHANNEL_DEFINE` and simulates a battery discharging from 4200 mV to 3300 mV.

Reviewer: ```suggestion

The sample includes a mock provider in the :file:`src/mock_battery_provider.c` file that registers a channel with :c:macro:`VTF_CHANNEL_DEFINE` and simulates a battery discharging from 4200 mV to 3300 mV.
```

This shows that a provider can use any data source, such as a fuel gauge library or a register read, as long as it implements the ``init()`` and ``sample()`` functions described in the subsystem documentation.

Configuration
*************

|config|

Devicetree configuration
========================

> The board overlay in :file:`boards/nrf7120dk_nrf7120_cpuapp.overlay` configures the following:

Reviewer: ```suggestion
The board overlay in the :file:`boards/nrf7120dk_nrf7120_cpuapp.overlay` file configures the following:
```

* ``nordic,vtf-region`` - Selects the SRAM region where ``vtf_snapshots`` are stored.
* ``nordic,vtf-die-temp-sensor`` - Selects the sensor used for die temperature capture.

The nRF7120 DK does not include a second temperature sensor.
The overlay therefore points ``nordic,vtf-die-temp-sensor`` to the same ``temp`` node used by the default fallback.
This demonstrates how to apply the devicetree override on a custom board without changing the reading on the DK.

On a custom board, point ``nordic,vtf-die-temp-sensor`` to the nodelabel of your own sensor.

.. _vtf_monitoring_sample_custom_battery:

Custom battery voltage backend

> ================================

Reviewer: ```suggestion
Custom battery voltage backend
==============================
```

.. code-block:: console

   west build -p -b nrf7120dk/nrf7120/cpuapp samples/vtf_monitoring -- -DCONFIG_VTF_BATTERY_VOLTAGE_MONITOR=n -DCONFIG_SAMPLE_VTF_CUSTOM_BATTERY_VOLTAGE_MONITOR=y

In this configuration, the sample logs a decreasing battery voltage until it resets to 4200 mV.

Building and running
********************

.. |sample path| replace:: :file:`samples/vtf_monitoring`

To build for the nRF7120 DK, use the ``nrf7120dk/nrf7120/cpuapp`` board target.
The following is an example of the CLI command:

.. code-block:: console

   west build -p -b nrf7120dk/nrf7120/cpuapp samples/vtf_monitoring

Testing
=======

|test_sample|

1. |connect_kit|
#. |connect_terminal|
#. Observe the periodic log output.
   With the default configuration, the output looks similar to the following example:

   .. code-block:: console

      [00:00:01.013,768] <inf> vtf_monitoring_sample: die temp: 25.00 C
      [00:00:01.013,784] <inf> vtf_monitoring_sample: battery voltage: 3300 mV

>       [00:00:01.013,790] <inf> vtf_monitoring_sample: XO offset: 100 ppm

Reviewer 3: set default XO offset to 0 ppm.
